### PR TITLE
[Digital Ocean] Fix sporadic volume detach error when volume is already detached

### PR DESCRIPTION
**Pull request by srikiz**
Based on our e2e tests, when the e2e tests are complete, we delete the cluster. As part of the delete workflow, we first detach the droplets from the volume, and then go about deleting the volume.

We found a sporadic error when volume is already detached to the droplet - https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/e2e-kops-do-calico/1412394584972988416. Looks like some kind of a race condition..

Error details below for quick reference.
```
droplet:253698717
	volume:8d266b3f-de59-11eb-81b7-0a58ac144312
volume:8d266b3f-de59-11eb-81b7-0a58ac144312	error deleting resources, will retry: failed to detach volume: 8d266b3f-de59-11eb-81b7-0a58ac144312, err: POST https://api.digitalocean.com/v2/volumes/8d266b3f-de59-11eb-81b7-0a58ac144312/actions: 404 (request "7ba87cb6-a105-4920-a073-d9e1325d527e") error detaching volume from droplet: volume is not attached to droplet
Not all resources deleted; waiting before reattempting deletion
	volume:8d266b3f-de59-11eb-81b7-0a58ac144312
	droplet:253698717
volume:8d266b3f-de59-11eb-81b7-0a58ac144312	error deleting resources, will retry: failed to detach volume: 8d266b3f-de59-11eb-81b7-0a58ac144312, err: POST https://api.digitalocean.com/v2/volumes/8d266b3f-de59-11eb-81b7-0a58ac144312/actions: 404 (request "a19dc990-3161-4e9d-be9f-fc86fd1cafb7") error detaching volume from droplet: volume is not attached to droplet
```

This PR first checks the response and check for `StatusNotFound` status before erroring out - I am not 100% sure this fixes, but wanted to put this and see if this helps. Please let me know in case you have any questions.

FYI - @timoreimann @rifelpet

### CODE REVIEW

File: pkg/resources/digitalocean/resources.go [hunk 1]
```diff
@@ -264,13 +265,16 @@ func deleteVolume(cloud fi.Cloud, t *resources.Resource) error {
...
+		action, resp, err := c.VolumeActionService().DetachByDropletID(context.TODO(), volume.ID, dropletID)
+		if resp != nil && resp.StatusCode != http.StatusNotFound {
+			if err != nil {
+				return fmt.Errorf("failed to detach volume: %s, err: %s", volume.ID, err)
+			}
+			if err := waitForDetach(c, action); err != nil {
+				return fmt.Errorf("error while waiting for volume %s to detach: %s", volume.ID, err)
+			}
 		}
+
```
Comment (hakman) [inline pkg/resources/digitalocean/resources.go:277]: Looks good. Maybe remove this line?
```suggestion
```

Comment (hakman) [reply pkg/resources/digitalocean/resources.go]: Please also squash the commits.

File: pkg/resources/digitalocean/resources.go [hunk 2]
```diff
@@ -264,12 +265,14 @@ func deleteVolume(cloud fi.Cloud, t *resources.Resource) error {
...
 	volume := t.Obj.(godo.Volume)
 	for _, dropletID := range volume.DropletIDs {
-		action, _, err := c.VolumeActionService().DetachByDropletID(context.TODO(), volume.ID, dropletID)
-		if err != nil {
-			return fmt.Errorf("failed to detach volume: %s, err: %s", volume.ID, err)
-		}
-		if err := waitForDetach(c, action); err != nil {
-			return fmt.Errorf("error while waiting for volume %s to detach: %s", volume.ID, err)
+		action, resp, err := c.VolumeActionService().DetachByDropletID(context.TODO(), volume.ID, dropletID)
+		if resp != nil && resp.StatusCode != http.StatusNotFound {
```
Comment (timoreimann) [inline pkg/resources/digitalocean/resources.go:269]: For the sake of readability and consistency, I'd first check if the error is non-nil and derive proper actions accordingly:

```go
if err != nil {
  if resp != nil && resp.StatusCode == http.StatusNotFound {
    // Volume is already detached, nothing to do.
    continue
  }
  return fmt.Errorf("failed to detach volume %s: %s", volume.ID, err)
}

if err := waitForDetach(c, action); err != nil {
  return fmt.Errorf("error while waiting for volume %s to detach: %s", volume.ID, err)
}
```

Comment (srikiz) [reply pkg/resources/digitalocean/resources.go]: Sure, updated.

**Review by hakman — APPROVED** [review]

**Pull request merged**